graph/telemetry/istio/appender/labeler.go

```diff
@@ -0,0 +1,118 @@
+package appender
+
+import (
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/graph"
+	"github.com/kiali/kiali/log"
+)
+
+const LabelerAppenderName = "labeler"
+
+// LabelerAppender is responsible for obtaining and attaching all k8s labels to all nodes in the graph.
+// Name: labeler
+type LabelerAppender struct{}
+
+// Name implements Appender
+func (f *LabelerAppender) Name() string {
+	return LabelerAppenderName
+}
+
+// IsFinalizer implements Appender
+func (a LabelerAppender) IsFinalizer() bool {
+	return true
+}
+
+// AppendGraph implements Appender
+func (f *LabelerAppender) AppendGraph(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo, _namespaceInfo *graph.AppenderNamespaceInfo) {
+	if len(trafficMap) == 0 {
+		return
+	}
+
+	labelNodes(trafficMap, globalInfo)
+}
+
+// labelNodes puts all k8s labels in the metadata for all nodes.
+func labelNodes(trafficMap graph.TrafficMap, gi *graph.AppenderGlobalInfo) {
+	// We need to know the names of the Istio labels for app and version because we do not label the nodes with those.
+	// There is no need to get the Istio label names multiple times, so get them once now.
+	istioLabelNames := config.Get().IstioLabels
+
+	for _, n := range trafficMap {
+		// can't get labels for nodes on the outside or inaccessible nodes, so just go to the next and ignore this one.
+		if b, ok := n.Metadata[graph.IsOutside]; ok && b.(bool) {
+			continue
+		}
+		if b, ok := n.Metadata[graph.IsInaccessible]; ok && b.(bool) {
+			continue
+		}
+
+		var labelsMetadata graph.LabelsMetadata
+
+		switch n.NodeType {
+		case graph.NodeTypeApp:
+			if n.Version != "" {
+				// the node is a "versioned-app" node
+				if wl, ok := getWorkload(n.Namespace, n.Workload, gi); ok {
+					labelsMetadata = copyMap(wl.Labels)
+				} else {
+					log.Debugf("Failed to obtain versioned-app details for [%+v]", n)
+				}
+			} else {
+				// getApp returns a list of workloads that make up the app. Capture the superset of all labels on all workloads.
+				// However, throw away any labels that have different values across the different workloads.
+				labelsMetadata = graph.LabelsMetadata{}
+				if app, ok := getApp(n.Namespace, n.App, gi); ok {
+					differentLabels := map[string]bool{}
+					for _, w := range app.Workloads {
+						if r, ok := getWorkload(n.Namespace, w.WorkloadName, gi); ok {
```

Comment:
Do you need to  `getWorkload`, is `w` not already the workload?

Reply:
`w` is a `WorkloadItem` and only contains minimal data:
![image](https://user-images.githubusercontent.com/2029470/156256288-ec8ed173-4234-45e9-957d-0f0ad12681f6.png)

Reply:
after calling that getWorkload(), this is the result (notice more data comes back in this, which is a WorkloadListItem:
![image](https://user-images.githubusercontent.com/2029470/156256430-b865d584-161b-48d4-8133-f803f87aff50.png)